### ISSUE TITLE
Restore automated Docker image publishing

### ISSUE THREAD

**Issue by sepek** [prosody/prosody-docker#72]
Most recent version of Prosody is 0.11.10 released on 2021-08-03 not available at https://hub.docker.com/r/prosody/prosody/
https://blog.prosody.im/prosody-0.11.10-released/  
Please provide an image for it.

**Comment by Zash** [comment]
We switched build / CI system and have yet to rig it up with Docker image creation. Sorry for the delay. Hopefully we will get to it before the next release.

**Comment by LeSpocky** [comment]
Meanwhile [0.11.12 is available](https://blog.prosody.im/prosody-0.11.12-released/) with an important security fix. @Zash, @mwild1 could you please look into this? 😇

**Comment by mwild1** [comment]
Yes, I'll try to look at it later this week. Unfortunately I have an important deadline that I'm working towards in the next couple of days.

A reminder that you should be able to easily use the Dockerfile here to build your own images. That's part of the goal of Docker :)

I'll post back to this issue with news soon, hopefully!

**Comment by LeSpocky** [comment]
> A reminder that you should be able to easily use the Dockerfile here to build your own images. That's part of the goal of Docker :)

Unfortunately, this is not so easy. _README.md_ says:

> **Note:** Using this Dockerfile outside of our build server is not a supported use-case.

The reason is you use a file *prosody.deb* which is not part of this repo. I guess it is generated by your build server?

**Comment by mwild1** [comment]
> The reason is you use a file prosody.deb which is not part of this repo. I guess it is generated by your build server?

Yes, the .deb file is the same one that's available in our package repository. Here's a direct link to the relevant .deb file for amd64 systems: https://packages.prosody.im/debian/pool/main/p/prosody/prosody_0.11.12-1~buster1_amd64.deb ([more here](https://packages.prosody.im/debian/pool/main/p/prosody/))

**Comment by zombiehoffa** [comment]
so this has been broken for a year and a whole major release cycle at this point? Is there any plan to get the docker images in docker hub updating again? Or are the updated ones being hosted elsewhere or something?

**Comment by mwild1** [comment]
Yes, it has, and yes, there is a desire to get them working again.

It's probably half a day of work, but I'm not finding much time for non-urgent infrastructure tasks right now. I've been working on authentication and account security improvements in Prosody thanks to a grant. If I don't meet milestones for that, I don't get paid, and if I don't get paid I can't pay my bills.

I understand it can be frustrating - it can be frustrating to me! But this is the reality of open-source. We don't have regular steady income to sustain the development of the project. The vast majority of the project activities happen on volunteered time, which is highly variable. Despite the drawbacks, we consider this better than many of the alternatives (Prosody Enterprise Edition, anyone?)

So, expect me to randomly find half a day free to work on this and be pleasantly surprised by a "fixed" notification at some point. But despite a desire to see this done, I can't provide any fixed date for when that might be able to happen.

Meanwhile, instructions for building images yourself are earlier in this thread.

Hope this helps explain the situation, and how such an issue can linger unresolved for so long.

**Comment by mwild1** [comment]
Funding for singular small tasks doesn't really work out for various reasons, primarily because it would be necessary to amass a lot of such tasks to allow a single developer to work and get paid for a month. Long-term funding opportunities such as grants and consulting contracts offer more stability.

In other words, if an open-source developer can choose between a grant or contract that would allow them to work on a range of issues for e.g. 6 months, or an offer of a half-day payment for a single task with no how to earn money through the rest of the month, it's obvious that the former is the only reasonable choice.

If you're really interested in supporting the project in some way, feel free to email developers@prosody.im and we can see what can be worked out. The project itself doesn't have any organization/company behind it, but individual developers may be up for paid work, if they are available.

**Comment by sepek** [comment]
So to give you a proper number: about 80.000€. That should suffice for paying a developer half a year.
Anyways, thanks for all your efforts. **I really appreciate it**. Maybe take the images offline as long as you can't provide new ones. It's fine to cut down features. And as long as this repository still exists. People can take care of it themselves.

**Comment by mwild1** [comment]
Hi folks,

I've been working on (re-)adding Docker outputs to our CI system. I have an image ready for testing :tada: 

The image `prosodyim/prosody:0.12` is multi-arch and based on debian:bullseye-slim with the prosody-0.12 package installed (this is the latest nightly build from our stable branch).

At this point I'm basically looking for a couple of people to give it a quick spin, and see if there are any obvious issues. If not, I'll add automated builds for 0.11 (since that's technically still a supported branch) and the latest stable release. The final images will be pushed to `prosody/prosody`. In the future we may also add additional image flavours, such as a tiny version based on alpine, for example.

The new images also support more environment variables for basic configuration than the old images did. I don't have time to document them right now, but I'll aim to do that this week. You can inspect the prosody.cfg.lua in the repo and container to get an idea of what's possible in the meantime.

Let me know how it goes!

**Comment by mwild1** [comment]
I just added some new test images, so now we have:

- `prosodyim/prosody:0.12`
- `prosodyim/prosody:0.11`
- `prosodyim/prosody:trunk`

These are only temporary, for testing, and will eventually migrate to the existing `prosody/prosody`.

I also added documentation of the environment variables: https://prosody.im/doc/docker#environment-variables

**Comment by Pinkbyte** [comment]
I confirm successful migration from prosody/prosody:0.11 to prosodyim/prosody:0.11 and later - to prosodyim/prosody:0.12 image. Plugins installation(new feature in 0.12 releases) are broken in 0.12 image, so keep that in mind. I failed to migrate to it from manually synced prosody-modules community repo. Some kind of problem with LuaRocks when trying to install plugin, i did not save logs, sorry.

Anyway, just my 2 cents that if something works for you in old 0.11 images it should work with new images too.

**Comment by mwild1** [comment]
Thanks for the feedback! (I haven't had any other so far).

I probably won't be able to look at this now until after FOSDEM in a few weeks, but I think at this point it will be a simple switch to push the new images live in place of the old ones.

**Comment by Zash** [comment]
```
~$ podman exec prosody prosodyctl install --server=https://modules.prosody.im/rocks/ mod_unified_push
Installing mod_unified_push in /var/lib/prosody/custom_plugins

Error: Failed finding Lua header files. You may need to install them or configure LUA_INCDIR.
Installing https://modules.prosody.im/rocks/mod_unified_push-13-1.src.rock
```

Looks like https://github.com/luarocks/luarocks/issues/1275

**Comment by poVoq** [comment]
When trying to run the :0.12 image with Podman (in rootful mode) I get the following error:

`usermod: UID '0' already exists`

And it fails to start.

**Comment by Zash** [comment]
@poVoq that looks like you are using the old images, not the new ones.

**Comment by poVoq** [comment]
no I was using `prosodyim/prosody:0.12`

Anyways, I switched to a dockerless system for Prosody now.

**Comment by Zash** [comment]
Huh, I thought the new images did not have this stuff

**Comment by merwan** [comment]
@mwild1 I migrated successfully from `prosody/prosody:0.11.5` image to `prosodyim/prosody:0.11` without doing any configuration change apart from changing the docker image tag.
Running the `about` command gives me this result:
```
Prosody 0.11 nightly build 143 (2022-01-24, 458c5f8d5d3e)
```
Which looks to be the version `0.11.13` of prosody.

I won't upgrade to v0.12 for the time being, but I will let you know when I do.
Thanks for providing this new image with the security fix for v0.11

**Comment by SkyBird233** [comment]
I have successfully migrated from a custom Docker image of Prosody (utilizing the official Prosody 0.12.0 DEB file) to a locally built version using files directly from the 'ng' folder of this repository. The Prosody version is `0.12 nightly build 209 (2023-08-30, a2ba3f06dcf4)`, and everything is working properly without any configuration changes.

I'm looking forward to the official image becoming available soon.

**Comment by MaximalCats** [comment]
Are there any updates on this? Last official release on dockerhub was over 3 years ago

**Comment by bhhaskin** [comment]
> Are there any updates on this? Last official release on dockerhub was over 3 years ago

4 years now. I don't think this is going to happen.

**Comment by LeSpocky** [comment]
> There still seems to be some progress on this, as a 13.0 tag was pushed 5 days ago to: https://hub.docker.com/r/prosodyim/prosody/tags

If I am not mistaken those images are not built from this Git repository.  So this here should be marked as deprecated or archived, and it should be made transparent where those other images come from.

**Comment by mwild1** [comment]
Actually both `prosody` and `prosodyim` are owned by the Prosody team. I mentioned the new org a few years ago in this very thread, here: https://github.com/prosody/prosody-docker/issues/72#issuecomment-1349113958

The images under 'prosodyim' are built automatically by our CI system. From the commit history, and the updates to the README in this repo earlier in the year, you will see that this repository is still in use ( https://github.com/prosody/prosody-docker/commits/master/ ).

It has been a hassle to get Docker Hub to approve us as an open-source project (they kept raising the bar, as they want most orgs to switch to a paid tier) and ultimately we want to self-host rather than pay Docker for hosting. We have a self-hosted registry set up (currently being tested by the Snikket project), and I expect we will move to that at some point.

As has been said earlier in the thread, if you want up-to-date prosody images, you can:

- Use the newer style CI images at https://hub.docker.com/r/prosodyim/prosody
- Build it yourself
- Use one of the third-party alternative images linked in the README

There are plenty of options.

**Comment by LeSpocky** [comment]
> It has been a hassle to get Docker Hub to approve us as an open-source project (they kept raising the bar, as they want most orgs to switch to a paid tier) and ultimately we want to self-host rather than pay Docker for hosting. We have a self-hosted registry set up (currently being tested by the Snikket project), and I expect we will move to that at some point.

Another option would be to let GitHub host those images through the ghcr.io registry. :wink:

**Comment by mwild1** [comment]
These "free" services subsidized by VCs and big tech come and go. We're in it for the long term - our Github usage is not :)

The only reason this Github org exists is because the person who first worked on docker images for Prosody was more comfortable working with it, and Docker Hub integrated with Github. It is likely we'll [transition it to Mercurial](https://hg.prosody.im/docker) along with our other code repositories, rather than increase our dependency on the goodwill of Microsoft.

**Comment by pixlxip** [comment]
> When trying to run the :0.12 image with Podman (in rootful mode) I get the following error:
> 
> `usermod: UID '0' already exists`
> 
> And it fails to start.

Hi, I got this error while trying to run in Docker, and was able to fix it by changing the permissions of the data directory to be owned by the id of the prosody user in the container:
```sh
chown 100:102 /path/to/data/
```
I'm not sure if this was best practice to do, but it seems to have worked.
